design/anova/adding_anova_to_mmrm_design_document.qmd

```diff
@@ -0,0 +1,157 @@
+
+If the user supplied `test = TRUE`, the model fits are subject to further tests to ensure they are nestable. If the data sets for the fits are not the same and the user supplied `refit = TRUE`, the data sets are harmonized and the models are refitted using the harmonized data set. The pairs of models then undergo likelihood ratio testing, and the results are augmented into the `tibble` of results.
+
+### Pseudocode
+
+```{r eval=FALSE}
+anova.mmrm <- function (object, ..., test = TRUE, refit = FALSE) {
+  
+  # If there are objects supplied to "..."
+  if (...length()) {
+    
+    # Gather all objects into one list
+    fits <- list(object, ...)
+    
+    # Ensure all the objects are mmrm fits
+    validate_mmrm_objects(fits)
+    
+    # Calculate the standard diagnostics no matter what
+    results <-
+      tibble(
+        call = lapply(fits, mmrm::component, "call"),
+        Model = row_number(),
+        df = vapply(fits, get_df, integer(1)), 
+        AIC = vapply(fits, AIC, numeric(1)),
+        BIC = vapply(fits, BIC, numeric(1)),
+        logLik = vapply(fits, logLik, numeric(1))
+      )
+    
+    # If the user has requested the likelihood ratio test...
+    if (test) {
+      
+      # Ensure all fits can be nested
+      ensure_same_formula_for_fixed_effects(fits)
+      ensure_same_time_group_subjectID_vars(fits)
+      ensure_nestable_covariance_structures(fits)
+      
+      # Handle differences in underlying data sets
+      data_sets_are_same <- compare_data_underlying_model_fits(fits)
+      if (!data_sets_are_same) {
+        if (refit) {
+          # Create one harmonized data set from all fits' underlying data
+          # Errors out if not possible
+          harmonized_data <- harmonize_underlying_data(fits)
+          
+          # Refit all models using harmonized data
+          fits <- lapply(fits, refit_mmrm, harmonized_data)
+        } else {
+          stop("Data underlying mmrm fits are not all the same and refit = FALSE")
+        }
+      }
+      
+      
+      results <- results |>
+        mutate(
+          # Label for the pair being compared (e.g., "3 vs 4")
+          test = c(NA, paste0(head(Model, -1), " vs ", tail(Model, -1))),
+          
+          # Calculate the likelihood ratio between the pairs using base::diff()
+          likelihood_ratio = c(NA, diff(logLik)),
+          
+          # Calculate the p-value using the ratios and differences in df
+          p_value = LR_pvalues(likelihood_ratio, c(NA, diff(df)))
+        )
+    }
+    
+  } else { 
+    
+    # If only one model fit was provided, calculate the significance of model
+    # terms, which is what ?stats::anova promises
+    out <- calculate_significance_of_model_terms(object)
+  }
+  
+  out
+}
+```
+
```

Comment:
Please start with prototype code here (just functions, without roxygen2 chunks or tests, but with text explanations to illustrate the flow etc.), please see other design docs as examples